src/bufr/BufrParser/Query/SubsetTable.h

```diff
@@ -11,25 +11,271 @@
 #include <string>
 #include <unordered_map>
 #include <set>
+#include <iostream>
+#include <sstream>
 
 #include "DataProvider/DataProvider.h"
 
 namespace Ingester {
 namespace bufr {
-    /// \brief Meta data for a query according to the BUFR subset table.
-    struct QueryData
+
+    class PathComponent;
+
+    /// \brief A node in the bufr data tree (metadata). This is a recursive data structure.
+    struct BufrNode : public std::enable_shared_from_this<BufrNode>
     {
-        bool isMissing;
-        int nodeId;
-        std::vector<std::string> pathComponents;
-        bool isString;
-        std::vector<size_t> seqPath;
-        std::vector<size_t> dimIdxs;
-        size_t idx;
-        bool requiresIdx;
+        std::weak_ptr<BufrNode> parent;
+        std::string mnemonic;
+        Typ type;
+        std::vector<std::shared_ptr<BufrNode>> children;
+        size_t nodeIdx;
+        size_t copyIdx;
+        bool hasDuplicates;
         TypeInfo typeInfo;
+
+        /// \brief Do this nodes child sequences appear as parts of the query string?
+        /// \return True if this node is a parent of a query path node.
+        bool isQueryPathParentNode()
+        {
+            return type == Typ::DelayedRep ||
+                   type == Typ::FixedRep ||
+                   type == Typ::DelayedRepStacked ||
+                   type == Typ::DelayedBinary;
+        }
+
+        /// \brief Does this node appear as part of the query string?
+        /// \return True if this node is a query path node.
+        bool isQueryPathNode()
+        {
+            bool isQueryPath = true;
+            if (!parent.expired())
+            {
+                isQueryPath = parent.lock()->isQueryPathParentNode();
+            }
+
+            return isQueryPath;
+        }
+
+        /// \brief Does this node add dimension to the resulting data? (i.e. is it a repeat node)
+        /// \return True if this node is a dimensioning node.
+        bool isDimensioningNode()
+        {
+            return type == Typ::DelayedRep ||
+                   type == Typ::FixedRep ||
+                   type == Typ::DelayedRepStacked ||
+                   type == Typ::Subset;
+        }
+
+        /// \brief Is this node a leaf node (i.e. a data node)
+        /// \return True if this node is a leaf node.
+        bool isLeaf() const
+        {
+            return type == Typ::Number ||
+                   type == Typ::Character;
+        }
+
+        /// \brief Does this type of node contain children?
+        /// \return True if this node is a container node.
+        bool isContainer() const
```

Comment:
What is an example of when this would happen and/or be needed?

Reply:
This function tells you if a node is a type of node that can contain children. It can be useful in a lot of ways, for example while recursively searching the BUFR subset tree you can use it to figure out if you need to dive deeper or not.